dotnet/src/SemanticKernel.Abstractions/Functions/KernelArguments.cs

```diff
@@ -65,7 +91,38 @@ public KernelArguments(IDictionary<string, object?> source, Dictionary<string, P
     /// <summary>
     /// Gets or sets the prompt execution settings.
     /// </summary>
-    public IReadOnlyDictionary<string, PromptExecutionSettings>? ExecutionSettings { get; set; }
+    /// <remarks>
+    /// The settings dictionary is keyed by the service ID, or <see cref="PromptExecutionSettings.DefaultServiceId"/> for the default execution settings.
+    /// When setting, the service id of each <see cref="PromptExecutionSettings"/> must match the key in the dictionary.
+    /// </remarks>
+    public IReadOnlyDictionary<string, PromptExecutionSettings>? ExecutionSettings
+    {
+        get => this._executionSettings;
+        set
+        {
+            // Clone the settings to avoid reference changes.
+            this._executionSettings = value is IDictionary<string, PromptExecutionSettings> dictionary
+                ? new Dictionary<string, PromptExecutionSettings>(dictionary)
+                : value?.ToDictionary(kv => kv.Key, kv => kv.Value);
```

Comment:
Having `IReadOnlyDictionary<string, PromptExecutionSettings>` as property type means that if user has custom implementation of `IReadOnlyDictionary<string, PromptExecutionSettings>`, it's also possible to set it here. But with this logic, when I will try to set `MySettingsDictionary` which implements both `IReadOnlyDictionary` and `IDictionary` or extends `Dictionary<string, PromptExecutionSettings>`, it will override it with `System.Collections.Generic.Dictionary`. 

```csharp
public class MySettingsDictionary : IDictionary<string, PromptExecutionSettings>, IReadOnlyDictionary<string, PromptExecutionSettings>
{
    // Implementation
}

[Fact]
public void Test()
{
    var mySettings = new MySettingsDictionary();
    var arguments = new KernelArguments() { ExecutionSettings = mySettings };

    Console.WriteLine(arguments.ExecutionSettings.GetType());
    // Before change: Namespace.MySettingsDictionary
    // After change: System.Collections.Generic.Dictionary
}
```

Not sure if it's a common case, but if somebody does this today, this change looks like behavioral breaking change. Even if nobody does this, this behavior doesn't look intuitive, because the property is abstracted, but it will always be `Dictionary`, so it makes sense to keep property type as `Dictionary<string, PromptExecutionSettings>`.

Reply:
The change was made to prevent reference changes after you set the `KernelArgument` settings, which prevents unexpected reference impact. 

That said this is an expected behavioral change where only break implementations that leverage an undesirable behavior.

The internal dictionary will still be exposed as `IReadOnlyDictionary` which keeps public API exposure also intact.

On your scenario this would only be a breaking change if the specialized Dictionary don't respect the Liskov Substitution Principle (Intentionally changing the expected Dictionary behavior) which also is a anti-pattern and ensuring code quality from the caller side.

Reply:
> On your scenario this would only be a breaking change if the specialized Dictionary don't respect the Liskov Substitution Principle (Intentionally changing the expected Dictionary behavior) which also is a anti-pattern and ensuring code quality from the caller side.

@RogerBarreto I think on our side we cannot be sure if specific SOLID principle or other pattern is respected on user side or not. But from the usage perspective, it may be confusing when I set specialized Dictionary (and I can do that, because public API allows me to do so), but it will be ignored and `System.Collections.Generic.Dictionary` will be used instead in all cases.

Imagine there is a scenario, when every time the new setting is added to the dictionary, user wants to verify if prompt execution settings are valid and log warning/throw exception if not. Today, it's possible to do that by implementing dictionary abstraction and set that extended implementation to `KernelArguments.ExecutionSettings`. But with new change, we will override it and user-defined custom validation and logging will be ignored. Just want to make sure if this sounds acceptable, maybe it's worth to communicate that through PR description and/or release notes.

In this case, it feels natural to convert `ExecutionSettings` property type to `Dictionary<string, PromptExecutionSettings>` (like in `PromptTemplateConfig.cs` for example), so we explicitly say that custom dictionary is not acceptable here, but we can't do that, because it will be a breaking change on compilation time. By modifying setter to always create `Dictionary`, this is breaking change in runtime, which may be even harder to troubleshoot. 

I understand the reason why it's implemented in this way, I just want to make sure that we understand the consequences.

Reply:
This is a double-edged sword, keep the original reference and change it or not. The only way I see to keep the original reference is to avoid auto-set `ServiceId` information and consider that value as optional for every case and not representing at all what is the current Key in the dictionary... not sure what impact this will have but seems to be a more simplistic approach... I will investigate that.

Reply:
@dmytrostruk After discussing with the team and valid considerations I did some changes to this PR and dropped the approach on `Auto-Setting` the `ServiceId` property internally which was the whole reason as well as the problem of `safe` copying/cloning. 

As a solution I just made this new property `Optional` allowing just to use it when provided and while keeping the reference as is.